<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="PITCh designer 2.1 is a free online tool to design gRNAs & homology arms for marker gene knock-in.">
    <meta name="author" content="Kazuki Nakamae">


    <title>PITCh KnockIn | About</title>
    <link rel="shortcut icon" href="images/SMGlogo.ico">


    <!-- Bootstrap Core CSS -->
    <link href="css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom CSS -->
    <link href="css/about.css" rel="stylesheet">

    <!-- Custom Fonts -->
    <link href="font-awesome/css/font-awesome.min.css" rel="stylesheet" type="text/css">
    <link href="https://fonts.googleapis.com/css?family=Lato:300,400,700,300italic,400italic,700italic" rel="stylesheet" type="text/css">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
        <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
        <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
    <![endif]-->

</head>

<body>

    <!-- Navigation -->
    <nav class="navbar navbar-default navbar-fixed-top topnav" role="navigation">
        <div class="container topnav">
            <!-- Brand and toggle get grouped for better mobile display -->
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                  <span class="sr-only">Toggle navigation</span>
                  <span class="icon-bar"></span>
                  <span class="icon-bar"></span>
                  <span class="icon-bar"></span>
              </button>
                <a class="navbar-brand topnav" href="./index.html">PITCh KnockIn</a>
            </div>
            <!-- Collect the nav links, forms, and other content for toggling -->
            <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
                <ul class="nav navbar-nav navbar-right">
                    <li>
                        <a href="./about.html">About</a>
                    </li>
                    <li>
                        <a href="./index.html">PITCh designer 2.1</a>
                    </li>
                    <li>
                        <a href="./contact.html">Contact</a>
                    </li>
                </ul>
            </div>
            <!-- /.navbar-collapse -->
        </div>
        <!-- /.container -->
    </nav>


    <!-- Header -->
    <a name="about"></a>
    <div class="body-pageheader img-responsive">
        <div class="container">
            <h1>PITCh KnockIn</h1>
            <p class="lead">
                An easy and efficient method for gene knock-in
            </p>
        </div>
    </div>
    <!-- /.body-pageheader -->

    <!-- Page Content -->

    <a name="services"></a>
    <div class="content-section-a">

        <div class="container" id="PITChintro">
            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <hr class="section-heading-spacer">
                    <div class="clearfix"></div>
                    <h2 class="section-heading">What are the PITCh systems ?</h2>
                    <p class="lead">
                        &ensp;The PITCh (Precise Integration into Target Chromosome) systems are the alternative methods of CRISPR-Cas9- and TALEN-mediated knock-in
                        <sup>
                         <a target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/26678082">1</a>
                       </sup> .
                        <br><br><br>
                    </p>
                    <h4 class="section-heading">Advantages</h4>
                    <p class="lead">
                        &ensp;These systems enable seamless gene knock-in with extremely short homologous sequences (10–40 bp), which can be easily added to the donor vector by PCR and<br> In-Fusion cloning
                        <sup>
                         <a target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/26678082">1</a>
                       </sup>.<br><br> &ensp;The PITCh knock-in is based on one of the repair mechanism of DNA double strand break,microhomology-mediated end-joining (MMEJ). Since MMEJ is independent on homologous recombination
                        (HR), which is mainly used for gene knock-in, PITCh system can be utilized in cells and organisms with low HR frequency.<br><br>
                    </p>
                    <h4 class="section-heading">Applications</h4>
                    <p class="lead">

                        &ensp;The PITCh systems have currently been applied in cultured cells (HEK293T, HeLa, HCT116, CHO), silkworms, frogs, zebrafish, and mice.<br>
                        <button type="button" class="btn btn-link btn-lg" data-toggle="modal" data-target=".bs-example-modal-lg">
                          Newly released examples >>>
                        </button>
                        <div class="modal fade bs-example-modal-lg" tabindex="-1" role="dialog" aria-labelledby="myLargeModalLabel">
                            <div class="modal-dialog modal-lg" role="document">
                                <div class="modal-content">
                                    <div class="modal-header">
                                        <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
                                        <h4 class="modal-title" id="myModalLabel">Newly released examples</h4>
                                    </div>
                                    <div class="modal-body">
                                        <ul>
                                            <li><a href="https://www.sciencedirect.com/science/article/pii/S1389172317310691">Kawabe, Yoshinori, et al. "Targeted knock-in of an scFv-Fc antibody gene into the hprt locus of Chinese hamster ovary cells using CRISPR/Cas9 and CRIS-PITCh systems." Journal of Bioscience and Bioengineering.(2018).</a></li>
                                            <li><a href="http://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1006784">Long, Quanxin, et al. "The role of host DNA ligases in hepadnavirus covalently closed circular DNA formation." PLoS Pathogens 13.12 (2017): e1006784.</a></li>
                                            <li><a href="http://www.ebiomedicine.com/article/S2352-3964(17)30213-X/fulltext">Yao, Xuan, et al. "CRISPR/Cas9–Mediated Precise Targeted Integration In Vivo Using a Double Cut Donor with Short Homology Arms." EBioMedicine (2017).</a></li>
                                            <li><a href="http://www.tandfonline.com/doi/full/10.1080/21655979.2017.1313645">Nakamae, Kazuki, et al. "Establishment of expanded and streamlined pipeline of PITCh knock-in–a web-based design tool for MMEJ-mediated gene knock-in, PITCh designer, and the variations of PITCh, PITCh-TG and PITCh-KIKO." Bioengineered 8.3 (2017): 302-308.</a></li>
                                            <li><a href="http://bio.biologists.org/content/6/5/706.long">Nakagawa, Yoshiko, et al. "Culture time of vitrified/warmed zygotes before microinjection affects the production efficiency of CRISPR-Cas9-mediated knock-in mice." Biology Open 6.5 (2017): 706-713.</a></li>
                                            <li><a href="https://www.nature.com/articles/nature20565">Suzuki, Keiichiro, et al. "In vivo genome editing via CRISPR/Cas9 mediated homology-independent targeted integration." Nature 540.7631 (2016): 144-149.</a></li>
                                            <li><a href="https://www.ncbi.nlm.nih.gov/pubmed/27894274">Aida, Tomomi, et al. "Gene cassette knock-in in mammalian cells and zygotes by enhanced MMEJ." BMC Genomics 17.1 (2016): 979.</a></li>
                                            <li><a href="http://www.nature.com/nprot/journal/v11/n1/abs/nprot.2015.140.html">Sakuma, Tetsushi, et al. "MMEJ-assisted gene knock-in using TALENs and CRISPR-Cas9 with the PITCh systems." Nature Protocols 11.1 (2016): 118-133.</a></li>
                                            <li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4632728/">Sakuma, Tetsushi, et al. "Homologous recombination-independent large gene cassette knock-in in CHO cells using TALEN and MMEJ-directed donor plasmids." International Journal of Molecular Sciences 16.10 (2015): 23849-23866.</a></li>
                                            <li><a href="http://www.nature.com/articles/srep08841">Hisano, Yu, et al. "Precise in-frame integration of exogenous DNA mediated by CRISPR/Cas9 system in zebrafish." Scientific Reports 5 (2015).</a></li>
                                            <li><a href="http://www.nature.com/articles/ncomms6560">Nakade, Shota, et al. "Microhomology-mediated end-joining-dependent integration of donor DNA in cells and animals using TALENs and CRISPR/Cas9." Nature Communications 5 (2014).</a></li>
                                        </ul>
                                    </div>
                                    <div class="modal-footer">
                                        <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                        <!-- /.modal fade bs-example-modal-lg -->
                    </p>
                </div>
                <div class="col-lg-5 col-lg-offset-2 col-sm-6">
                    <br><br><br><br><br><br>
                    <img class="img-responsive img-rounded" src="images/PITCh.jpeg" alt="">
                </div>
            </div>

        </div>
        <!-- /.container -->

    </div>
    <!-- /.content-section-a -->

    <div class="content-section-b">

        <div class="container">

            <div class="row">
                <div class="col-lg-5 col-lg-offset-1 col-sm-push-6  col-sm-6">
                    <hr class="section-heading-spacer">
                    <div class="clearfix"></div>
                    <h2 class="section-heading">Web Application : <br>"<a target="_blank" href="./index.html">PITCh designer 2.x</a>"</h2>
                    <p class="lead">
                        <span class="strong">
                      PITCh designer 2.x is a free online tool to design<br>
                      sgRNA and homology arms for gene knock-in.</span>
                        <br><br> &ensp;This tool <span class="strong">only needs</span> sequence around the target region. <span class="strong">After the sequence submission, followed by the selection of the target base, the application automatically designs sgRNA, microhomologies, and primers for constructing a donor vector for PITCh knock-in and for genotyping</span>.<br><br>
                        <a target="_blank" href="./index.html">Go to the PITCh designer 2.x >>></a>
                    </p>
                    <br>
                    <h2>Procedure</h2>
                    <h3>1. Paste sequence :</h3>
                    <p class="lead">
                        <br> &ensp;
                        <span class="strong">Paste the FASTA-formatted sequence</span> around the target region, and click submit.<br><br>
                    </p>
                    [NOTICE]
                    <ul>
                        <li>The acceptable length of the sequence is 100-15,000[bp].</li>
                        <li><span class="strong">The target base should not be located on either end of the sequence.</span><br>
                            <span class="strong">The tool requires 48-bp flanking sequences on the both sides of the target base regardless of the lengths of microhomologies.</span></li>
                    </ul>
                </div>
                <div class="col-lg-5 col-sm-pull-6  col-sm-6">
                    <a href="./index.html">
                    PITCh designer 2.x :
                    <img class="img-responsive" src="images/overview.png" alt="">
                  </a>
                    <br><br> 1. Paste sequence :
                    <img class="img-responsive" src="images/form.jpg" alt="">
                </div>
            </div>

        </div>
        <!-- /.container -->

    </div>
    <!-- /.content-section-b -->

    <div class="content-section-c">

        <div class="container">

            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <div class="clearfix"></div>
                    <h4 class="section-heading">Options :</h4>
                    <p class="lead">
                        &ensp;You can use design options for the PITCh knock-in.
                        <br><br>
                        <a class="btn btn-default btn-lg" role="button" data-toggle="collapse" href="#collapseExample" aria-expanded="false" aria-controls="collapseExample">
                            <span class='glyphicon glyphicon-chevron-down'></span>details
                        </a>
                    </p>
                </div>
                <div class="col-lg-5 col-lg-offset-2 col-sm-6">
                    <img class="img-responsive" src="images/option.png" alt=""><br><br>
                </div>
            </div>

        </div>
        <!-- /.container -->

        <div class="collapse" id="collapseExample">
            <div class="content-section-c-collapse">

                <div class="container">
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">1 : Reading frame</h4>
                            &ensp;Three reading frames or 'No frame' can be selected (as shown in the right figure).<br><br> Options :
                            <ul>
                                <li>Frame 1 (default)</li>
                                <li>Frame 2</li>
                                <li>Frame 3</li>
                                <li>No frame</li>
                            </ul>
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/ReadingFrame.jpeg">
                                <img class="media-object img-response" src="images/ReadingFrame.jpeg" alt="...">
                            </a>
                        </div>
                        <hr class="section-heading-spacer">
                    </div>
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">2 : Adjustment of reading frame</h4>
                            &ensp;PITCh designer 2.x offers two methods to prevent frameshifting.<br><br> Options :
                            <ul>
                                <li>5'C-insertion (default)<br><br> C-insertion mode fills in cytosine base(s) to adjust the reading frame.<br> The number of fill-in bases depends on the selected frame and the position of knock-in.</li>
                            </ul>
                            <ul>
                                <li>Codon deletion<br><br> Codon deletion mode removes extra base(s).</li>
                            </ul>
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/MHmethods.jpeg" target="_blank">
                                <img class="media-object img-response" src="images/MHmethods.jpeg" alt="...">
                            </a>
                        </div>
                        <hr class="section-heading-spacer">
                    </div>
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">3 : Knock-in cassette</h4>
                            &ensp;The tool designs the primers for the construction of a targeting vector to knock-in the following two preset cassettes or user-specifyed sequence</user-specifyed>.<br><br> Options :
                            <ul>
                                <li>EGFP2APuroR (default)</li><br><br> Usage :
                                <ul>
                                    <li>C-terminal tagging</li>
                                    <a href="https://www.ncbi.nlm.nih.gov/pubmed/26678082" target="_blank">
                            (Sakuma et al, 2015)
                          </a>
                                </ul>
                                <br>
                            </ul>
                            <ul>
                                <li>CMV-EGFP2APuroR-polyA</li><br><br> Usage :
                                <ul>
                                    <li>Transgenesis or knock-in/knock-out</li>
                                    <a href="http://www.tandfonline.com/doi/full/10.1080/21655979.2017.1313645" target="_blank">
                                        (Nakamae et al, 2017)
                                      </a>
                                </ul>
                                <br>
                            </ul>
                            <ul>
                                <li>User defined insert <img src="images/new_version2.0.png" alt=""></li>
                            </ul>
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/GOI.png" target="_blank">
                                <img class="media-object img-response" src="images/GOI.png" alt="...">
                            </a>
                        </div>
                        <br>
                        <h5>The plasmids for EGFP2APuroR knock-in are available from <a href="https://www.addgene.org/browse/article/16395/" target="_blank">Addgene</a>:<br>
                            <a href="https://www.addgene.org/browse/article/16395/" target="_blank">https://www.addgene.org/browse/article/16395/</a></h5>
                        <hr class="section-heading-spacer">
                    </div>
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">4-6 : Microhomology lengths and PAM sequence requirement</h4>
                            <br>
                            <h5>4,6 : Lengths of left/right microhomologies</h5>
                            &ensp;The acceptable length of each microhomology arm is 10-40[bp]<br> (The default length is 40 bp.)<br><br>

                            <h5>5 : PAM sequence requirement</h5>
                            &ensp;PAM sequence can be defined using IUB codes.<br> (The default sequence is 5'-NGG-3' ; SpCas9)<br><br> *PITCh system using Cpf1 is unconfirmed.
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/gRNAandMH.jpeg" target="_blank">
                                <img class="media-object img-response" src="images/gRNAandMH.jpeg" alt="...">
                            </a>
                        </div>
                        <div class="row">
                            <div class="col-md-6">
                                IUB codes
                                <sup>
                        <a target="_blank" href="https://pmgf.osu.edu/sites/pmgf.osu.edu/files/imagecache/IUB_Nucleotide_codes.pdf">2</a>
                      </sup>
                                <table class="table table-striped table-center">
                                    <thead>
                                        <tr>
                                            <th>Code</th>
                                            <th>Definition</th>
                                            <th>Mnemonic</th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        <tr>
                                            <th scope="row">G</th>
                                            <td>G</td>
                                            <td><span class="Mnemonic-strong">G</span>uanine</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">A</th>
                                            <td>A</td>
                                            <td><span class="Mnemonic-strong">A</span>denine</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">T</th>
                                            <td>T</td>
                                            <td><span class="Mnemonic-strong">T</span>hymine</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">C</th>
                                            <td>C</td>
                                            <td><span class="Mnemonic-strong">C</span>ytosine</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">R</th>
                                            <td>A, G</td>
                                            <td>pu<span class="Mnemonic-strong">R</span>ines</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">Y</th>
                                            <td>C, T</td>
                                            <td>p<span class="Mnemonic-strong">Y</span>rimidines</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">M</th>
                                            <td>A, C</td>
                                            <td>a<span class="Mnemonic-strong">M</span>ino</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">K</th>
                                            <td>G, T</td>
                                            <td><span class="Mnemonic-strong">K</span>eto</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">S</th>
                                            <td>C, G</td>
                                            <td><span class="Mnemonic-strong">S</span>trong</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">W</th>
                                            <td>A, T</td>
                                            <td><span class="Mnemonic-strong">W</span>eak</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">H</th>
                                            <td>A, C, T</td>
                                            <td>not G</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">B</th>
                                            <td>C, G, T</td>
                                            <td>not A</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">V</th>
                                            <td>A, C, G</td>
                                            <td>not T</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">D</th>
                                            <td>A, G, T</td>
                                            <td>not C</td>
                                        </tr>
                                        <tr>
                                            <th scope="row">N</th>
                                            <td>A, C, G, T</td>
                                            <td>a<span class="Mnemonic-strong">N</span>y</td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                            <!-- /.col-md-6 -->
                            <div class="col-md-6">
                                PAM sequences
                                <sup>
                        <a target="_blank" href="https://www.addgene.org/crispr/guide/">3</a>
                      </sup>
                                <table class="table table-center">
                                    <thead>
                                        <tr>
                                            <th>Species</th>
                                            <th>PAM sequence ( 5' -> 3' )</th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        <tr>
                                            <th scope="row"><span class="scientificname">Streptococcus pyogenes</span> (SP) ; SpCas9</th>
                                            <td>NGG</td>
                                        </tr>
                                        <tr>
                                            <th scope="row"><span class="scientificname">Staphylococcus aureus</span> (SA) ; SaCas9</th>
                                            <td>NNGRRT or NNGRR(N)</td>
                                        </tr>
                                        <tr>
                                            <th scope="row"><span class="scientificname">Neisseria meningitidis</span> (NM) ; NmCas9</th>
                                            <td>NNNNGATT</td>
                                        </tr>
                                        <tr>
                                            <th scope="row"><span class="scientificname">Streptococcus thermophilus</span> (ST) ; St1Cas9</th>
                                            <td>NNAGAAW</td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                            <!-- /.col-md-6 -->
                        </div>
                        <!-- /.row -->
                        <hr class="section-heading-spacer">
                    </div>
                    <!-- /media-->
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">7 : Genomes of target organism <img src="images/new_version2.0.png" alt=""></h4>
                            &ensp;The tool can show <a href="http://crispr.dbcls.jp/" target="_blank">CRISPRdirect</a> or
                            <a href="https://gggenome.dbcls.jp/en/" target="_blank">GGGenome</a> URL link to check sgRNA and primer specificity against genomic DNA, if you choose genome sequence.
                            <br> &ensp;<a href="http://crispr.dbcls.jp/" target="_blank">CRISPRdirect</a> and <a href="https://gggenome.dbcls.jp/en/" target="_blank">GGGenome</a> perform exhaustive searches against genomic sequence. Approxmately 600 genomes
                            are recorded in these tools.
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/genome_flowchart.png" target="_blank">
                                <img class="media-object img-response" src="images/genome_flowchart.png" alt="...">
                            </a>
                        </div>
                        <hr class="section-heading-spacer">
                    </div>
                    <!-- /media-->
                    <div class="media">
                        <div class="media-body">
                            <h4 class="media-heading">8 : Parameters of genotyping primers <img src="images/new_version2.0.png" alt=""></h4>
                            &ensp;The tool helps the users to set parameters (Tm or GC%) for the primers and automatically designs primers for genotyping.
                            <br><br> &ensp;This automated tool designs the primers according to the following procedure.
                            <br><br>
                            <ol>
                                <li>Perform exhaustive primer design in the distance of 100-200 bp from the corresponding positions.</li>
                                <li>Select primers according to the user's setting.</li>
                                <li>The optimized primers are chosen using rankings based on range of parameters and primer length.</li>
                                *The longer primers have a higher priority in the primers with the same Tm and GC%.
                            </ol>
                        </div>
                        <div class="media-right media-middle">
                            <a href="./images/primerdesign.png" target="_blank">
                                <img class="media-object img-response" src="images/primerdesign.png" alt="...">
                            </a>
                        </div>
                        <br>
                        <hr class="section-heading-spacer">
                    </div>
                    <!-- /media-->
                </div>
                <!-- /container-->
            </div>
            <!-- /.content-section-c -->
        </div>
        <!-- /.collapse -->
    </div>
    <!-- /.content-section-c -->

    <div class="content-section-b">

        <div class="container">

            <div class="row">
                <div class="col-lg-5 col-lg-offset-1 col-sm-push-6  col-sm-6">
                    <div class="clearfix"></div>
                    <h3>2. Select target :</h3>
                    <p class="lead">
                        &ensp;The application displays nucleotide buttons, which are color-coded according to their targetability and molecular species.<br>
                    </p>
                    <table class="table">
                        <thead>
                            <tr>
                                <th>Color</th>
                                <th>Targetability</th>
                                <th>Base</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <th scope="row">Green</th>
                                <td><span class='glyphicon glyphicon-ok ok-color'></span></td>
                                <td>Adenine</td>
                            </tr>
                            <tr>
                                <th scope="row">Red</th>
                                <td><span class='glyphicon glyphicon-ok ok-color'></span></td>
                                <td>Thymine</td>
                            </tr>
                            <tr>
                                <th scope="row">Blue</th>
                                <td><span class='glyphicon glyphicon-ok ok-color'></span></td>
                                <td>Guanine</td>
                            </tr>
                            <tr>
                                <th scope="row">Yellow</th>
                                <td><span class='glyphicon glyphicon-ok ok-color'></span></td>
                                <td>Cytocine</td>
                            </tr>
                            <tr>
                                <th scope="row">Light Blue</th>
                                <td><span class='glyphicon glyphicon-remove notok-color'></span></td>
                                <td>Any</td>
                            </tr>
                            <tr>
                                <th scope="row">Gray</th>
                                <td>Out of range</span>
                                </td>
                                <td>Any</td>
                            </tr>
                        </tbody>
                    </table>
                    <p class="lead">
                        &ensp;After selecting the desired position, the webtool begins to design the sequences for PITCh knock-in.
                    </p>
                </div>
                <div class="col-lg-5 col-sm-pull-6  col-sm-6">
                    2. Select target:
                    <img class="img-responsive" src="images/Selection.jpeg" alt="">
                </div>
            </div>
            <!-- /.row -->
            <div class="row">
                <div class="col-lg-5 col-lg-offset-1 col-sm-push-6  col-sm-6">
                    <h3>3. Show result :</h3>
                    <p class="lead">
                        &ensp;The design results consist of <span class="strong">sequence information</span> and <span class="strong">graphical images of knock-in and donor vector construction</span> ( as shown in the left figure ).
                    </p>
                    Downloadable items
                    <ul>
                        <li>Summary</li>
                        <ul>
                            <li>The designed sequences are exported to tab-delimited text or CSV format, which can be viewed in Microsoft Excel.</li>
                        </ul><br>
                        <li>Whole sequence of donor vector <img src="images/new_version2.0.png" alt=""></li>
                        <li>Graphical image of knock-in</li>
                        <li>Graphical image of donor vector construction</li>
                    </ul>
                    </p>
                </div>
                <div class="col-lg-5 col-sm-pull-6  col-sm-6">
                    3. Show result :
                    <img class="img-responsive" src="images/Result.jpeg" alt="">
                </div>
            </div>
            <!-- /.row -->

        </div>
        <!-- /.container -->

    </div>
    <!-- /.content-section-b -->


    <div class="content-section-c">

        <div class="container">

            <!-- hidden
            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <h3>New function from version 2.x</h3>
                </div>
            </div>
            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <h3>Protocol generator</h3>
                    <p class="lead">
                        &ensp;Result form can export the PDF of protocol.
                        <br>&ensp;This procedure is updated to reflect current design including sequences and their Tm value.
                        <br>&ensp;This document consist of:
                        <ol>
                            <li>Construction of the CRIS-PITCh(v2) donor vector.</li>
                            <li>Construction of the CRISPR-Cas9 vector for CRIS-PITCh (v2).</li>
                            <li>Genotyping by Sequencing.</li>
                        </ol>
                    </p>
                    <br>
                </div>
                <div class="col-lg-5 col-lg-offset-2 col-sm-6">
                    Protocol generator :
                    <img class="img-responsive" src="images/protocolgenerator.png" alt="">
                </div>
            </div>
            -->
            <!-- /.row -->
            <br>
            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <h3>Primer viewer <img src="images/new_version2.0.png" alt=""></h3>
                    <p class="lead">
                        &ensp;The PITCh designer 2.x can not only automatically select best primer pair for genotyping but help users to manually choose it on this viewer ( as shown in the right figure ).
                        <br>&ensp;This viewer can show primers for left junction / right junction / out-out PCR. By clicking on the arrow button on the left button form, you can change the primer type.
                        <br><br>&ensp;After clicking the left button for selecting primer pair, the viewer displays sequence information at the bottom of the graphical map. The sequence information can also be download by clicking the "DOWNLOAD SEQENCE"
                        button.
                    </p>
                </div>
                <div class="col-lg-5 col-lg-offset-2 col-sm-6">
                    Primer viewer :
                    <img class="img-responsive" src="images/primerviewer.png" alt="">
                </div>
            </div>
            <!-- /.row -->
            <br>
            <div class="row">
                <div class="col-lg-5 col-sm-6">
                    <h3>Efficiency annotation <img src="images/new_version2.1.png" alt=""></h3>
                    <p class="lead">
                        &ensp;The PITCh designer 2.1 can analyze the knock-in design and adds "efficiency annotations" to the knock-in design. We showed that the group of annotated knock-in design significantly achieved the high efficiency of precise knock-in compared with the no annotation group in HEK293T cells ( as shown in the right figure ).
                        <br>&ensp;The sequence features, which contribute to the gain of precise knock-in efficiency, will be shown in Desirable labels. On the other hand, Undesirable labels indicate the feature causing less knock-in efficiency.
                        <br><br>&ensp;The features were determined based on our previous research (<a href="https://www.biorxiv.org/content/10.1101/2022.06.27.496697v1">Nakamae et al., 2022</a>). The PITCh designer searches sequence context resulting in the mutation patterns affecting the knock-in efficiency.
                    </p>
                </div>
                <div class="col-lg-5 col-lg-offset-2 col-sm-6">
                    Efficiency annotation:
                    <img class="img-responsive" src="images/design_annotation.png" alt="">
                    <img class="img-responsive" src="images/annotation_stat.png" alt="">
                </div>
            </div>
            <!-- /.row -->
        </div>
        <!-- /.container -->

    </div>
    <!-- /.content-section-c -->


    <a name="contact"></a>
    <div class="banner">

        <div class="container">

            <div class="row">
                <div class="col-lg-6">
                    <h2>Let's design PITCh !!!</h2>
                </div>
                <div class="col-lg-6">
                    <ul class="list-inline banner-social-buttons">
                        <li>
                            <a href="./index.html" class="btn btn-default btn-lg"><i class="glyphicon glyphicon-hand-right"></i> <span class="network-name">Go to the PITCh designer 2.x</span></a>
                        </li>
                    </ul>
                </div>
            </div>

        </div>
        <!-- /.container -->

    </div>
    <!-- /.banner -->

    <!-- Footer -->
    <footer>
        <div class="container">
            <div class="row">
                <div class="col-lg-12">
                    <ul class="list-inline">
                        <li>
                            <a href="./about.html">About</a>
                        </li>
                        <li class="footer-menu-divider">&sdot;</li>
                        <li>
                            <a href="./index.html">PITCh designer 2.1</a>
                        </li>
                        <li class="footer-menu-divider">&sdot;</li>
                        <li>
                            <a href="./contact.html">Contact</a>
                        </li>
                    </ul>
                    <br>
                    <hr>
                    <br>
                    <p class=s>
                        <a href='./index.html'>PITCh designer 2.1</a> |
                        <a href='https://smg.hiroshima-u.ac.jp'>
                            Molecular Genetics Laboratory, Graduate School of Integrated Sciences for Life, Hiroshima University, Japan
                  		</a><br> This page is licensed under a
                        <a target='_blank' href='http://creativecommons.org/licenses/by/2.1/jp/deed.en'>
                  		Creative Commons Attribution 2.1 Japan License</a>.
                    </p>
                </div>
            </div>
        </div>
    </footer>

    <!-- jQuery -->
    <script src="js/jquery-1.11.3.js"></script>

    <!-- Bootstrap Core JavaScript -->
    <script src="js/bootstrap.min.js"></script>
    <!-- GoogleAnalytics -->
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');

        ga('create', 'UA-96900517-1', 'auto');
        ga('send', 'pageview');
    </script>
</body>

</html>